### PR TITLE
Plumb IdP configuration at KDB side

### PR DESCRIPTION
**Pull request by abbra**
#   KDB: support IdP configuration

 When IdP configuration is provided, take it into account:
    
  - idp-specific Kerberos ticket policy would be applied
    
  - Presence of IdP link in a Kerberos principal would cause KDB to request the KDC to first ask RADIUS server for an initial challenge and then supply the response in the communication with the client side of OTP pre-authentication as an OTP token to use
    
The latter requires additional changes from MIT Kerberos side.
    
Fixes: https://pagure.io/freeipa/issue/8824

# External IdP: add idpUser object to be able to link to IdP
    
A link to IdP configuration needs to be stored in a user object. Handle it similarly to RADIUS proxy configuration link

### CODE REVIEW

File: install/updates/40-idp.update [hunk 1]
```diff
@@ -2,3 +2,12 @@ dn: cn=idp,$SUFFIX
...
 default: cn: idp
+
+dn: cn=ipaidpconfiglink,cn=index,cn=userRoot,cn=ldbm database,cn=plugins,cn=config
+only:cn: ipaidpconfiglink
+default:objectClass: nsIndex
+default:objectClass: top
+default:nsSystemIndex: false
+add:nsIndexType: eq
+add:nsIndexType: pres
+add:nsIndexType: sub
```
Comment (flo-renaud) [inline install/updates/40-idp.update:13]: We can also add `referint-membership-attr:ipaidpconfiglink` to install/updates/25-referint.update

File: daemons/ipa-kdb/ipa_kdb_principals.c [hunk 2]
```diff
@@ -321,6 +322,27 @@ static void ipadb_validate_radius(struct ipadb_context *ipactx,
...
+static void ipadb_validate_idp(struct ipadb_context *ipactx,
+                               LDAPMessage *lentry,
+                               enum ipadb_user_auth *ua)
+{
+    struct berval **vals;
+
+    if (!(*ua & IPADB_USER_AUTH_IDP))
+        return;
+
+    /* Ensure that the user has a link to a RADIUS config. */
```
Comment (flo-renaud) [inline daemons/ipa-kdb/ipa_kdb_principals.c:334]: Typo: RADIUS -> IdP

**Review by flo-renaud — COMMENTED** [review]
Hi @abbra,
the changes on the api LGTM. 
Not comfortable enough with the IPA KDB driver but seems good also.

File: daemons/ipa-kdb/ipa_kdb_principals.c (hunk 2 above)
Comment (abbra) [reply daemons/ipa-kdb/ipa_kdb_principals.c:334]: Thanks, fixed.

File: install/updates/40-idp.update (hunk 1 above)
Comment (abbra) [reply install/updates/40-idp.update:13]: Good note, added.

**Comment by abbra** [comment]
I'll add MIT Kerberos build with Pavel's changes next week and we'll be able to create a test that goes through the whole pipeline.

**Comment by abbra** [comment]
I need to adopt few tests to see the presence of `ipaidpconfiglink` attribute. Will work on that next week.

**Pull request closed without merging**